src/cs50/sql.py

```diff
@@ -316,8 +319,14 @@ def shutdown_session(exception=None):
                 # Join tokens into statement, abbreviating binary data as <class 'bytes'>
                 _statement = "".join([str(bytes) if token.ttype == sqlparse.tokens.Other else str(token) for token in tokens])
 
+                # If COMMIT or ROLLBACK, turn on autocommit mode
+                if command in ["COMMIT", "ROLLBACK"] and "TO" not in statement:
+                    if not self._in_transaction:
+                        raise RuntimeError("transactions must be initiated with BEGIN or START TRANSACTION")
+                    self._in_transaction = False
+
                 # Execute statement
-                result = connection.execute(sqlalchemy.text(statement))
+                result = self._session.execute(sqlalchemy.text(statement))
```

Comment:
Should we call `self._session.commit()` down below if we're executing `COMMIT` here when we're in a transaction?

Reply:
Re: using the methods instead of passing to the database, the main difference in behavior is that the `commit` method will [clear loaded database state](https://docs.sqlalchemy.org/en/13/orm/session_api.html#sqlalchemy.orm.session.Session.commit), so the next execution will load the database anew. The `rollback` method has no differences in the [documentation](https://docs.sqlalchemy.org/en/13/orm/session_api.html#sqlalchemy.orm.session.Session.rollback) compared to just executing the statement.

I'm not sure adding several conditions (if command is begin, if command is rollback, if command is commit) is worth any perceived stylistic benefit of using the methods here. So I think -- since only `commit` has any real behavioral difference compared to just executing -- it would make sense to not execute any `COMMIT` statements passed by the user. This also resolves your sub-question here, since only the `commit()` execution would happen. It would also make sense to not execute `BEGIN`, since the transaction will be started automatically by SQLAlchemy. In contrast, `ROLLBACK` can take many forms, and there's seemingly no real functionality difference, so it probably makes sense to just execute the user's query.

Proposed solution:

```python
if command not in ["BEGIN", "COMMIT"]:
    result = self._session.execute(sqlalchemy.text(statement))```

Reply:
Allow me to let @kzidane respond here, but on a related note... If this `BEGIN` or `START` `statement` proves to be invalid, yielding an error of some sort when `execute` is called, do we reset `_in_transaction` to `False` and re-enable autocommit before raising an exception of our own?

Reply:
Hmm. So currently, no -- we have our own exception if the user tries to execute `BEGIN` but is already in a transaction. But in the current code any problem with the actual execution won't reset `_in_transaction`. That said, if we go forward with the code change noted earlier in this thread, then those problems ultimately won't be raised in the first place because the query won't be executed.

Reply:
How so? What if a student has incorrect syntax like `START TRANSACTION WHERE 1`? That wouldn't err until `execute` is called, no?

Reply:
Yes, that's right. Currently if the student has incorrect syntax then `execute` is where the error will be thrown. But if we decide to let SQLAlchemy and the dbapis handle the automatic starting of transactions (as in the code excerpt above), then that means we will also be preventing the execution of `BEGIN|START|COMMIT`. So any invalid syntax _won't_ be raised there. From a pedagogical perspective I think there may be a point to just executing the user's queries, period, if only to allow the user to learn from mistakes in syntax. And if it means we execute the user's `COMMIT` and then also call the method `commit` (so as to clear the database state as well) then it may be worth the educational benefit, even though we end up committing an empty transaction.

Reply:
Just to be clear, then, what's the downside of passing BEGIN, START TRANSACTION, COMMIT, and/or ROLLBACK to `execute`?

Reply:
I wasn't able to find any reasons explicitly against directly executing the commands. Maybe @kzidane has some reasoning that I haven't thought of. In the case of COMMIT, the SQLAlchemy session has a function `commit` that additionally clears the database state (so it gets updated data on next execution), but from what I've read in the [SQLAlchemy docs](https://docs.sqlalchemy.org/en/13/orm/session_api.html) the BEGIN/START and ROLLBACK commands are the same as calling their respective methods on the session. So only COMMIT has a "downside" but that's easily corrected by just calling the `commit` method as well.

Reply:
I was thinking it just sounds redundant to execute `BEGIN|COMMT|ROLLBACK` if `BEGIN` is executed automatically and `Session.commit()` and `Session.rolback()` execute `COMMIT` and `ROLLBACK` respectively and clean up state. Perhaps it's cleaner to just remember the state we need to remember when we see a `BEGIN` (i.e., that we're in a transaction) and simply discard these queries, relying on the method calls instead?

Reply:
I can see the reasoning here, but this would add several lines of code (if command is rollback then rollback, if command is BEGIN|ROLLBACK|COMMIT don't execute) and if the user had invalid syntax no error would ever be raised and they wouldn't be able to learn from that. That's not a trade-off I think is worth the trouble but I'll of course do whatever is preferred. Just let me know.

Reply:
I do think we should catch any syntax errors in students' statements if possible!

Reply:
Ok then will keep it simple, execute anyway so that any syntax errors are caught!